### DISCUSSION TITLE
sr_module_oper_data_update timeout management

### DISCUSSION THREAD

**Discussion by NLako-6WIND** [sysrepo/sysrepo#3387] [Ideas]
Hi,

Context:
I've encountered some issues where when a whole module (/*) was dependent on operational get subscriptions (each subpart had its own subscription), and we had a significant timeout used for rpc-timeout in Netopeer2. If the provider wasn't able to provide the data, Netopeer2 was blocked until the timeout. Worst, even if the provider wanted to gracefully stop and unsubscribe, it could not, as the shm_sub->oper_get_lock was READ taken by Netopeer2 -> dying in agony by timing out on each individual subscription.

From the first point of view, using timeout_ms for each subscription seems OK, as if 1 timed out, we would return an error without trying to process other subscriptions. But, if each subscription took something like 1/3 of timeout_ms and we had over X amount of subscriptions, we would already (+-) surpass the timeout_ms value without notifying a timeout to the requester.

My idea is to use a fraction of timeout_ms each time, something like:
`global_sub_timeout = timeout_ms / mod->shm_mod->oper_get_sub_count`
and
`used_timeout = global_sub_timeout / set->count`

What are your thoughts on this?

**Comment by michalvasko** [comment]
What does this solve exactly? I realized that there is this issue and decided to keep the timeout meaning as previously meaning of a subscription, not the whole API call.

**Reply by NLako-6WIND** [reply]
In my specific case, this change addressed an issue where a requester was prematurely closing sockets due to its use of the same timeout value for both internal operations and external requests. netopeer2/sysrepo might exceed this timeout if the data provider didn't respond promptly. While not a critical problem overall, and adjustable in both sysrepo and the requester, I wanted to highlight this potential interaction.
Regardless, thank you very much for your reply.
